File: solutions/102. Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        q=deque([root])
        res=[]
        node=root
        if not root:
            return 
        while len(q)>0:
            vals=[]
            for i in range(len(q)):
        
                node=q.popleft()
                vals.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(vals)
        return res
                
        
        
        
        
        
        
        
        
        
        
        
        # copy answer
        '''
        if not root: return []
        queue, res = deque([root]), []
        
        while queue:
            cur_level, size = [], len(queue)
            for i in range(size):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                cur_level.append(node.val)
            res.append(cur_level)
        return res
        '''
        
                
            

File: solutions/102. Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        
        //dfs
        
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        levelHelper(res, root, 0);
        return res;
    }
    
    public void levelHelper(List<List<Integer>> res, TreeNode root, int height) {
        if (root == null) return;
        if (height >= res.size()) {
            res.add(new LinkedList<Integer>());
        }
        res.get(height).add(root.val);
        System.out.println(res);
        levelHelper(res, root.left, height+1);
        levelHelper(res, root.right, height+1);}
    
    
       
        /*
        Queue <TreeNode> q= new LinkedList<>();
        List<List<Integer>> res =new ArrayList<>();
        q.add(root);
        
        if (root==null) {return res;}
        
        while (!q.isEmpty() ) {
            List<Integer> vals = new ArrayList<Integer>();
            int n=q.size();
            for (int i=0; i<n;i++)  {
